ludwig/features/sequence_feature.py

```diff
@@ -248,29 +250,108 @@ def update_metrics(self, targets, predictions):
+            if training:
+                return self._logits_training(inputs, target, training)
+            else:
+                return self._logits_prediction(inputs)
+        else:
+            # Tagger Decoder
+            if training:
+                return self.decoder_obj(inputs)
+            else:
+                return self._logits_prediction(inputs)
+
+
+    def _logits_training(self, inputs, target, training=None):
+        input = inputs['hidden'] # shape [batch_size, seq_size, state_size]
+        encoder_end_state = self._prepare_decoder_input_state(inputs)
 
-    def predictions(
+        logits = self.decoder_obj.decoder_training(
+            input,
+            target=target,
+            encoder_end_state=encoder_end_state
+        )
+        return logits  # shape = [b, s, c]
+
+    def _logits_prediction(self, inputs):
+        return inputs
+
+    def predictions(self, inputs, training=None):
+
+        # Generator Decoder
+        if training:
+            return self._predictions_training(inputs, training=training)
+        else:
+            return self._predictions_prediction(inputs, training=training)
+
+    # todo tf2 need to determine if the section of code is needed
+    def _predictions_training(self, inputs, training=None):    # not executed
+        # inputs == logits
+        probs = softmax(inputs)
+        preds = tf.argmax(inputs)
+        return {'predictions': preds, 'probabilities': probs}
+
+    def _predictions_prediction(
             self,
-            inputs   # logits
+            inputs,   # encoder_output, encoder_output_state
+            training=None
     ):
-
-        logits = inputs[LOGITS]
+        if isinstance(self.decoder_obj, SequenceGeneratorDecoder):
+            encoder_output = inputs[LOGITS]['hidden'] # shape [batch_size, seq_size, state_size]
+            # form dependent on cell_type
+            # lstm: list([batch_size, state_size], [batch_size, state_size])
+            # rnn, gru: [batch_size, state_size]
+            encoder_output_state = self._prepare_decoder_input_state(inputs[LOGITS])
+
+            logits = self.decoder_obj.decoder_inference(
+                encoder_output,
+                encoder_end_state=encoder_output_state,
+                training=training
+            )
+        else:
+            # Tagger Decoder  todo tf2 reconcile tensor shape for inputs[LOGITS][HIDDEN]
+            logits = self.decoder_obj(inputs[LOGITS], training=training)
 
         probabilities = tf.nn.softmax(
             logits,
             name='probabilities_{}'.format(self.name)
         )
-        predictions = tf.argmax(
-            logits,
-            -1,
-            name='predictions_{}'.format(self.name),
-            output_type=tf.int64
-        )
+
+        if isinstance(self.decoder_obj, SequenceGeneratorDecoder) and \
+                self.decoder_obj.beam_width > 1:
+            predictions = self.decoder_obj.decoder_beam_search(
```

Comment:
this looks to me like, if we have the beam search, we are decoding twise, once to obtain logits and once to obtain predictions with beam search (the if at 320 is true).
If I'm understanding correctly, this is wrong. We should call either the regular or the beam decoding procedure, collect from both logits and predictions, compute the probabilities, and if predictions are None (like the tagger would do) than also compute the argmax.

Reply:
You are correct.  I observed in this note https://github.com/uber/ludwig/pull/699#issuecomment-633767935
>From what I can tell, BeamSearchDecoder does not have an obvious way to caclulate logits. This means to create the collection logits, predictions and probabilities will require both tfa.seq2seq.BeamSearchDecoder and BasicDecoder classes.
> 
> Regardless of beam_width setting, I can use BasicDecoder to calcuate logits and probabilties. If beam_width > 1, then use BeamSearchDecoder with the outputs.samp_ids data structure for predictions, otherwise predictions are calculated using the argmax() function on logits.

In looking at the `BeamSearchDecoder` in the debugger, I could not find the equivalent of the `BasicDecoder` `outputs.rnn_output` structure to reconstruct the the `logits` tensor.  This is why I ended up using of the two decoders:
* `BasicDecoder` to get `logits` tensor from `outputs.rnn_output`
* `BeamSearchDecoder` to   `predictions` from the beam search `outputs.predicted_id`

I would have preferred to just just use `BasicDecoder` if `beam_width == 1` and `BeamSearchDecocder` if `beam_width > 1`.   

I'm most likely missing something.  I'm open to suggestion to simplify the logic.